lib/posix/fs.c

```diff
@@ -0,0 +1,399 @@
+
+	if (posix_fs_get_ptr(fd, (void **)&ptr, false)) {
+		set_pthread_errno(EBADF);
+		return -1;
+	}
+
+	rc = fs_close(ptr);
+
+	/* Free file ptr memory */
+	posix_fs_free_fd(fd);
+
+	if (rc < 0) {
+		set_pthread_errno(-rc);
+		return -1;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Write to a file.
+ *
+ * See IEEE 1003.1
+ */
+ssize_t write(int fd, char *buffer, unsigned int count)
+{
+	ssize_t rc;
+	struct fs_file_t *ptr = NULL;
+
+	if (posix_fs_get_ptr(fd, (void **)&ptr, false)) {
+		set_pthread_errno(EBADF);
+		return -1;
+	}
+
+	rc = fs_write(ptr, buffer, count);
+	if (rc < 0) {
+		set_pthread_errno(-rc);
+		return -1;
+	}
+
+	return rc;
+}
+
+/**
+ * @brief Read from a file.
+ *
+ * See IEEE 1003.1
+ */
+ssize_t read(int fd, char *buffer, unsigned int count)
+{
+	ssize_t rc;
+	struct fs_file_t *ptr = NULL;
+
+	if (posix_fs_get_ptr(fd, (void **)&ptr, false)) {
+		set_pthread_errno(EBADF);
+		return -1;
+	}
+
+	rc = fs_read(ptr, buffer, count);
+	if (rc < 0) {
+		set_pthread_errno(-rc);
+		return -1;
+	}
+
+	return rc;
+}
+
+/**
+ * @brief Move read/write file offset.
+ *
+ * See IEEE 1003.1
+ */
+int lseek(int fd, int offset, int whence)
+{
+	int rc;
+	struct fs_file_t *ptr = NULL;
+
+	if (posix_fs_get_ptr(fd, (void **)&ptr, false)) {
+		set_pthread_errno(EBADF);
+		return -1;
+	}
+
+	rc = fs_seek(ptr, offset, whence);
+	if (rc < 0) {
+		set_pthread_errno(-rc);
+		return -1;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Open a directory stream.
+ *
+ * See IEEE 1003.1
+ */
+DIR *opendir(const char *dirname)
+{
+	int rc, fd;
+	struct fs_dir_t *ptr = NULL;
+
+	fd = posix_fs_alloc_fd((void **)&ptr, true);
+	if ((fd < 0) || (ptr == NULL)) {
+		set_pthread_errno(EMFILE);
+		return NULL;
+	}
+	memset(ptr, 0, sizeof(struct fs_dir_t));
+
+	rc = fs_opendir(ptr, dirname);
+	if (rc < 0) {
+		posix_fs_free_fd(fd);
+		set_pthread_errno(-rc);
+		return NULL;
+	}
+
+	return ptr;
+}
+
+/**
+ * @brief Close a directory stream.
+ *
+ * See IEEE 1003.1
+ */
+int closedir(DIR *dirp)
+{
+	int rc;
+
+	if (dirp == NULL) {
+		set_pthread_errno(EBADF);
+		return -1;
+	}
+
+	rc = fs_closedir(dirp);
+
+	/* Free file ptr memory */
+	posix_fs_free_ptr(dirp);
+
+	if (rc < 0) {
+		set_pthread_errno(-rc);
+		return -1;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Read a directory.
+ *
+ * See IEEE 1003.1
+ */
+struct dirent *readdir(DIR *dirp)
+{
+	int rc;
+
+	if (dirp == NULL) {
+		set_pthread_errno(EBADF);
+		return NULL;
+	}
+
+	rc = fs_readdir(dirp, &fdirent);
+	if (rc < 0) {
+		set_pthread_errno(-rc);
+		return NULL;
+	}
+
+	memcpy(pdirent.d_name, fdirent.name, MAX_FILE_NAME);
+	return &pdirent;
+}
+
+/**
+ * @brief Rename a file.
+ *
+ * See IEEE 1003.1
+ */
+int rename(const char *old, const char *new)
+{
+	int rc;
+
+	rc = fs_rename(old, new);
+	if (rc < 0) {
+		set_pthread_errno(-rc);
+		return -1;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Remove a directory entry.
+ *
+ * See IEEE 1003.1
+ */
+int unlink(const char *path)
+{
+	int rc;
+
+	rc = fs_unlink(path);
```

Comment:
If the file is open, it can't be removed from the filesystem.  File descriptor allocation must have a reference count, and unlink() should only set a flag, that's checked during the time a file is closed, and only then it's passed to the filesystem.  Unless, of course, there's just one reference, in which case you can remove the file at that moment.

Reply:
As of now, this is not being handled and the assumption is application will do the due diligence of proper file open and close sequence.